receiver/prometheusreceiver/internal/metadata.go

```diff
@@ -30,31 +33,90 @@ type MetadataService interface {
 type MetadataCache interface {
 	Metadata(metricName string) (scrape.MetricMetadata, bool)
 	SharedLabels() labels.Labels
+	LoadObservedValue(key string) (*ObservedValue, bool)
+	StoreObservedValue(key string, ts int64, value float64)
+	LastScrapeTime() int64
+	EvictUnused()
 }
 
 type mService struct {
-	sm *scrape.Manager
+	sm    *scrape.Manager
+	cache sync.Map
 }
 
-func (t *mService) Get(job, instance string) (MetadataCache, error) {
-	targetGroup, ok := t.sm.TargetsAll()[job]
+func (ms *mService) Get(job, instance string) (MetadataCache, error) {
+	cacheKey := job + "|" + instance
+	c, ok := ms.cache.Load(cacheKey)
+	if ok {
+		mc := c.(*mCache)
+		mc.accessTimes++
+		return mc, nil
+	}
+
+	targetGroup, ok := ms.sm.TargetsAll()[job]
 	if !ok {
 		return nil, errors.New("unable to find a target group with job=" + job)
 	}
 
-	// from the same targetGroup, instance is not going to be duplicated
+	// from the same targetGroup, instance won't be duplicated
 	for _, target := range targetGroup {
 		if target.DiscoveredLabels().Get(model.AddressLabel) == instance {
-			return &mCache{target}, nil
+			mc := &mCache{t: target, startValues: make(map[string]*ObservedValue), accessTimes: 1}
+			ms.cache.Store(cacheKey, mc)
+			return mc, nil
 		}
 	}
 
 	return nil, errors.New("unable to find a target with job=" + job + ", and instance=" + instance)
 }
 
-// adapter to get metadata from scrape.Target
+// ObservedValue is used to store the start time, first observed value and last seen value of cumulative data
+type ObservedValue struct {
+	Ts            int64
+	StartValue    float64
+	LastSeen      float64
+	lastAccessRun int64
+}
+
+var dummyObservedValue = &ObservedValue{}
+
+// a structure that caches a scrape target and some other metadata related to the target
 type mCache struct {
-	t *scrape.Target
+	t           *scrape.Target
+	startValues map[string]*ObservedValue
+	accessTimes int64
+}
+
+// EvictUnused is used at the end of each scrape run for removing cached startValues which are not accessed in the past
+// maxUnusedRuns, as metrics can come and go from a scrape target endpoint
+func (m *mCache) EvictUnused() {
```

Comment:
How does it evict if the target is down and never comes up?

Reply:
@rghetia even if the target is down, or not in good health, prometheus still try to initialize an appender and feed data to it (with Value equals to NaN), as I commented on in the rewritten EndToEnd test.  logic can be found from: https://github.com/prometheus/prometheus/blob/851131b0740be7291b98f295567a97f32fffc655/scrape/scrape.go#L933-L935

however, I did make an assumption that a configured target will never be evicted even if it's down, during the lifecycle of opencensus agent.  the eviction is for metrics come and go, such as container metrics from a cadvisor endpoint

Reply:
SG.